### DISCUSSION TITLE
constraint cost limit exceeded - tolerance range?

### DISCUSSION THREAD

**Discussion by DV-Bongingi** [SESMG/SESMG#162] [Q&A] [answered]
I've set a specific constraint cost limit in the energysystem tab. This value got exceeded by around 6 %. Is there a tolerance range around the constraint cost limit for the solver an if so, how is it defined?
E. g.: set limit= 8500 tCO2e/a --> result constraint costs= 8990 tCO2e/a.

Thanks for your consideration and kind regards
Dominik

**Comment by GregorBecker** [comment, marked as answer]
Dear @DV-Bongingi, 

please reproduce the following two commits: 
- https://github.com/SESMG/SESMG/commit/8bf5176df56c24e324ebdff1256249294f9bbd31
- https://github.com/SESMG/SESMG/commit/1265ea6ab423749c51bfff9c972d819a8005c4ed

I hope this will solve the discrepancy in your emission balance. 

If not feel free to contact us again.

**Reply by DV-Bongingi** [reply]
Dear @GregorBecker ,
thanks a lot, from the first try this seems to work perfectly fine.
Constraint costs are now right on point. Thanks for fast response and commits!

**Reply by jtock** [reply]
Hi @DV-Bongingi,
perfect, that it works for the version you use!
The problem occurred while setting constraint cost limits and using storages in the system. You need to be aware that model results of previous runs (if you made some) can also be affected. They are maybe not comparable anymore to results you generate after the bugfix. Maybe you should check if this is the case and rerun just to be sure.
Greetings Jan!

**Reply by DV-Bongingi** [reply]
Hi @jtock ,
thanks for advise, yes I needed to rerun previous scenarios, but its does work without problems.
As far as I understand things, the SESMG got restructured and developed since the release of the version I use (0.2.0 from Aug 2022). I saw that it's now possible to do a pareto calculation with e. g. emissions and monetary costs. I really would like to use these functions, as they could possibly answer questions of my study. I'm unsure though, wether I would still be able to use my input spreadsheet that I've just finished. 
Can someone give me advise on how to switch to new version and what needs to be considered or even how to run both versions in different virtual environments? I'm afraid of deleting anything that might lead to non-working SESMG when processing on my own.
We could meet online if that is accelerating the process.
Many thanks
Dominik

**Reply by jtock** [reply]
Hi @DV-Bongingi,
you are right that we added some new features to the SESMG. We also solved some bugs and increased the quality of the results. 
Your first question: you will still be able to use the version you have installed right now. There will be a second version on your machine with its own virtual environment if you download a newer version.
We changed the input sheet and the input values which are necessary to run a model on several times. As far as i know your project,  you will have to add new weather data since we changed the way to model pv systems. We probably also changed some value names which effect your project. Due to the changes in the calculation I assume that the results of version 0.4.0 will differ from the ones you already have.
You can take a look in the actual model_definition_example to evaluate the effort you will have to make. You will find it here: https://github.com/SESMG/SESMG/blob/master/model_definition_example.xlsx .
Contact us if you want to install the actual version, since we have an installation issue on windows pcs right now. The actual version is under heavy development right now.
Greetings Jan

**Reply by DV-Bongingi** [reply]
Hi @jtock,
thanks for the informations. I had a look at the example sheet, it's version 0.3.0 available for download there. Adding GHI and DNI shouldn't be the problem, but it looks like the TwoInputTransformer is not integrated in this version as there are no columns for input2. I necessarily need this transformer type to do my calculations. Am I right in assuming that this means I can't use the actual version without huge effort? Is there maybe another way to use the pareto-optimization within my version?
Kind regards
Dominik

**Reply by jtock** [reply]
The version of the model_definition_example is already updated to 0.4.0. We just forgot to change it in the first sheet.
You are right that it is not implemented so far in the new version. I need to consider with the development team if there is a way to implement it without too much afford. I can give you an answer on that next week. (Sorry, that I will not be able to help you earlier!)
Greetings Jan

**Reply by jtock** [reply]
Hi @DV-Bongingi,
sadly we have no capacity to implement it in the code right now. You can try to implement it by yourself. It should be possible to implement GenericTwoInputTransformer you are looking for. But to be honst, I can not estimate the work which is necessary to do it. 
You can compare the older versions and the new one and try to reproduce it step by step. Maybe you start looking for it in the components.py.
Greetings Jan

**Reply by jtock** [reply]
Hi @DV-Bongingi, it should be possible to use your two-input/two-output transformer. You need to add the missing columns to the spreadsheet of the actual version. Please make sure to use the same terms / names as in the spreadsheet of your version.
We would thank you if you could also add it to the example spreadsheet, after you tested the changes on your system. (https://github.com/SESMG/SESMG/blob/master/model_definition_example.xlsx) 
Greetings! Jan

**Reply by DV-Bongingi** [reply]
Thanks @jtock and team for implementing, I will let you know as soon as I was able to test transformation and when everything works fine, adding to example is no problem.

**Reply by chrklemm** [reply]
@DV-Bongingi please try to restart the installation from scartch in a completely new folder. The error message indicates, that there was a problem during the implementation of the virtual environment, in which the SESMG is installed. Make sure, that you only reference two digits of your python version (e.g., "3.8" instead of "3.8.3") when starting the installation file.

**Reply by GregorBecker** [reply]
> Hi @jtock,
> I downloaded v0.5.0 and installed python, graphviz and cbc according to the installation manual.
> When trying to run SESMG I receive the error message "activate.bat" not found. Reading the "Run_SESMG_for_windows.cmd" batchfile tells me, the programm is looking for the directory "cd Scripts/start /b activate.bat", but in the downloaded structure of SESMG-ZIP folder, there is no such folder named "Scripts" on the level of the commandfile.
> Can you give me advise, how to solve the issue?
> (by the way: in the older version 0.2.0 the folder "Scripts" does exist at the described directory)
> Greetings

Can you move this thread to a new discussion so that it is clear that this is a new topic and the error is findable to others with the same one?
